### PR TITLE
[SPARK-15706] [SQL] Fix Wrong Answer when using IF NOT EXISTS in INSERT OVERWRITE for DYNAMIC PARTITION

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

`IF NOT EXISTS` in `INSERT OVERWRITE` should not support dynamic partitions. If we specify `IF NOT EXISTS`, the inserted statement is not shown in the table. 

This PR is to issue an exception in this case, just like what Hive does. Also issue an exception if users specify `IF NOT EXISTS` if users do not specify any `PARTITION` specification. 
#### How was this patch tested?

Added test cases into `PlanParserSuite` and `InsertIntoHiveTableSuite`

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @liancheng @hvanhovell @yhuai

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -171,6 +171,17 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
     val tableIdent = visitTableIdentifier(ctx.tableIdentifier)
     val partitionKeys = Option(ctx.partitionSpec).map(visitPartitionSpec).getOrElse(Map.empty)
 
+    if (ctx.EXISTS != null && ctx.partitionSpec == null) {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: We could enforce this in grammar.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -171,6 +171,17 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with Logging {
...
+      throw new ParseException(
+        s"IF NOT EXISTS is not expected when Partition Spec is not specified.", ctx)
+    }
+
+    val dynamicPartitionKeys = partitionKeys.filter { case (k, v) => v.isEmpty }
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: NIT: `partitionKeys.filter(_2.isEmpty)` is a bit shorter.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Sure, will do it. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: Yeah! Thank you!

**Comment by hvanhovell** [comment]
@gatorsmile the overall approach seems good to me. We are currently fixing this for the SQL codepath. I was wondering if are other codepaths that can cause this unwanted behavior? If there are then we should move the check to the `InsertInto` class.

**Comment by gatorsmile** [comment]
I think your concern is valid. Will add an `assert` in `InsertIntoTable`.

So far, dynamic partitioning is used by the `insertInto` API. However, there is no way to specify `IF NOT EXISTS` in the other API `insertInto`. 
```scala
data.write.mode(SaveMode.Overwrite).partitionBy("part").insertInto("partitioned")
```
The default value is always `false`, as shown below: https://github.com/apache/spark/blob/a71d1364ae87aa388128da34dd0b9b02ff85e458/sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala#L454

However, there still exist other bugs in `insertInto`. Will submit it later.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
@hvanhovell Does the latest code changes resolve your comment? Thanks!

**Comment by yhuai** [comment]
Seems the behavior of the following query is weird (if it works). (the case without IF NOT EXISTS) 
```
INSERT OVERWRITE TABLE table_with_partition
partition (p1='a',p2)
SELECT 'blarr3'
```

For this case, the number of column (1) of the query is less than the required number of columns (2). There should be an exception, right?

**Comment by gatorsmile** [comment]
@yhuai You are right. This is not a good test case for verifying this. I will add a case like 
```scala
      sql(
        """
          |INSERT OVERWRITE TABLE table_with_partition
          |partition (p1='a',p2) IF NOT EXISTS
          |SELECT 'blarr3', 'newPartition'
        """.stripMargin)
```

Actually, the above statement returns an empty result set, which is still a wrong answer.

Regarding your follow-up question, I also realize this is another bug we should capture, as shown above: https://github.com/apache/spark/pull/13447#issuecomment-223409590. My original plan is to submit a PR for addressing that issue after https://github.com/apache/spark/pull/12313 is merged. If https://github.com/apache/spark/pull/12313 will not be merged in this release, I will do it now. Thanks!

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @hvanhovell @yhuai Could you please review this PR again? Thanks!

**Comment by yhuai** [comment]
Thanks. LGTM. Merging to master and branch 2.0.

**Pull request closed without merging**